### DISCUSSION TITLE
Including an empty router to a FastAPI app, an later on attaching other routers to the initial router will still endup with an app with no routes defined

### DISCUSSION THREAD

**Discussion by Mumbawa** [fastapi/fastapi#12737] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from fastapi import APIRouter

app = FastAPI()

router1 = APIRouter(prefix="/router1") 

# router1 is included in the app here (at this poiny it's empty (it has no
# routes defined nor included routers))
app.include_router(router1) 

router2 = APIRouter(prefix="/router2")
router3 = APIRouter(prefix="/router3")

@router2.get("/hello")
async def hello2():
    return {"Hello": "World"}

@router2.get("/goodbye")
async def goodbye2():
    return {"Goodbye": "World"}

@router3.get("/hello")
async def hello3():
    return {"Hello": "World"}

@router3.get("/goodbye")
async def goodbye3():
    return {"Goodbye": "World"}

# Attaching router2 and router3 to router1
router1.include_router(router2)
router1.include_router(router3)

# WITHOUT line below THE APP WILL END UP WITH NO ROUTES!!!!!
# In other words I have to include router1 in the app AGAIN if
# I want end up with routes in the app, Which is weird since I already
# included router1 in the app above.

# app.include_router(router1) 

if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app, host="127.0.0.1", port=8000)
```

### Description

The problem is that if I add a router to a fastapi app (let's call it my_router), and later on I add routes (or attach other routers) to  "my_router", the app itself will end up with no routes! 

The only way to endup with routes at all in the app is by attaching AGAIN my_router to the app. Which I think it's weird or probably a bug

### Operating System

Linux

### FastAPI Version

0.115.3

### Pydantic Version

2.9.2

### Python Version

3.12.7

**Comment by sinisaos** [comment]
@Mumbawa I think if you include an empty router in the global app object, later changes (including other routers) have no effect because the app object takes precedence. As [noted in the docs](https://fastapi.tiangolo.com/reference/apirouter/#fastapi.APIRouter.include_router--example), you can include an `APIRouter` inside another `APIRouter`, but for this to take effect you must include those routers in the global app object **at the end**. You just need to remove first `app.include_router(router1)` and uncomment the last one `app.include_router(router1)`. I hope that makes sense.

**Reply by Mumbawa** [reply]
I do not understand why **_" later changes (including other routers) have no effect because the app object takes precedence"_** would prevent that routes added later on  to already-included routers end up included in the fastapi app.

I know that removing the first `app.include_router(router1)` and only use the last one it will work, however the code that I provided was just a simple example that I made after finding this wierd behavior in a larger app.

Here why I consider that this is problematic:

I have an application where I am defining endpoints **at runtime**... At some point you know that a router should be included in the app... This router may be empty or may contain some routes already... Later on I realize that I need to add new routes to this router... So one would expect that once these new routes are added to the router they will end up included in the app itself... Why would I have to include the router in the app again?

**Reply by sinisaos** [reply]
> I do not understand why **_" later changes (including other routers) have no effect because the app object takes precedence"_**

Sorry, I got it wrong at first, but I find this section in the docs that explains the current behavior. Quotes from docs
"APIRouters are not "mounted", they are not isolated from the rest of the application.

This is because we want to include their path operations in the OpenAPI schema and user interfaces.

Since we can't just isolate them and "mount" them independently of the others, path operations are "cloned" (re-created) instead of being included directly."

There is also a section in the docs why we need to include [all the routers before including them in the app](https://fastapi.tiangolo.com/tutorial/bigger-applications/#include-an-apirouter-in-another).

You can try using `app.mount` like this
```python
from fastapi import FastAPI

app = FastAPI()

router1 = FastAPI()
app.mount(app=router1, path="/router1")

router2 = FastAPI()
router3 = FastAPI()


@router2.get("/hello")
async def hello2():
    return {"Hello": "World"}


@router2.get("/goodbye")
async def goodbye2():
    return {"Goodbye": "World"}


@router3.get("/hello")
async def hello3():
    return {"Hello": "World"}


@router3.get("/goodbye")
async def goodbye3():
    return {"Goodbye": "World"}


# mounting router2 and router3 to router1
router1.mount(path="/router2", app=router2)
router1.mount(path="/router3", app=router3)


if __name__ == "__main__":
    import uvicorn

    uvicorn.run(app, host="127.0.0.1", port=8000)
```
but that's you lose the Openapi `docs`. `curl` works
```bash
> curl http://localhost:8000/router1/router3/hello
{"Hello":"World"}
> curl http://localhost:8000/router1/router3/goodbye
{"Goodbye":"World"}
```

**Comment by YuriiMotov** [comment, marked as answer]
That's how FastAPI works - when you use `.include_router()` it goes through all sub-routers and paths of the router you passed as a parameter and includes all paths it can find. So, if at this point router is not configured yet, it will not find any sub-routers and paths.

Just follow the rule "Configure router and only then include it into another router" and everything will work.

```
# main.py

from fastapi import FastAPI

from my_app.router1 import router1

app = FastAPI()

app.include_router(router1) 

if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app, host="127.0.0.1", port=8000)
```

```
# router1.py
from fastapi import APIRouter

from my_app.router2 import router2
from my_app.router3 import router3
from my_app.settings import config


router1 = APIRouter(prefix="/router1")

# Attaching router2 and router3 to router1 if needed
if config.include_router_2:
    router1.include_router(router2)
if config.include_router_3:
    router1.include_router(router3)
```

```
# router2.py
from fastapi import APIRouter

router2 = APIRouter(prefix="/router2")

@router2.get("/hello")
async def hello2():
    return {"Hello": "World"}

@router2.get("/goodbye")
async def goodbye2():
    return {"Goodbye": "World"}
```

```
# router3.py
from fastapi import APIRouter

router3 = APIRouter(prefix="/router3")

@router3.get("/hello")
async def hello3():
    return {"Hello": "World"}

@router3.get("/goodbye")
async def goodbye3():
    return {"Goodbye": "World"}
```
